contracts/ValidateSPV.sol

```diff
@@ -15,187 +15,106 @@ library ValidateSPV {
-        (_nOutputs, _outputs) = extractAllOutputs(_tx);
-        if (keccak256(_nOutputs) == keccak256(hex'00')) { 
-            return (_nInputs, _inputs, _nOutputs, _outputs, _locktime, _txid);
-        }
-
-        // Extract locktime
-        _locktime = _tx.extractLocktimeLE();
-
-        // Calculate txid
-        _txid = transactionHash(_version, _nInputs, _inputs, _nOutputs, _outputs, _locktime);
-    }
-
-    /// @notice         Parses tx input
-    /// @param _tx      The raw byte tx
-    /// @return         Raw bytes num inputs and inputs
-    function extractAllInputs(bytes memory _tx) public pure returns (bytes memory _nInputs, bytes memory _inputs) {
-        _nInputs = _tx.extractNumInputsBytes();
-        uint8 _tmpN = _tx.extractNumInputs();
-
-        for (uint8 i = 0; i < _tmpN; i++) {
-            _inputs = _inputs.concat(_tx.extractInputAtIndex(i));
+    ) internal pure returns (bool) {
+        // Shortcut the empty-block case
+        if (_txid == _merkleRoot && _index == 0 && _intermediateNodes.length == 0) {
+            return true;
         }
-    }
-
-    /// @notice         Parses tx outputs
-    /// @param _tx      The raw byte tx
-    /// @return         Raw bytes num outputs and outputs
-    function extractAllOutputs(bytes memory _tx) public pure returns (bytes memory _nOutputs, bytes memory _outputs) {
-        _nOutputs = _tx.extractNumOutputsBytes();
-        uint8 _tmpN = _tx.extractNumOutputs();
 
-        for (uint8 i = 0; i < _tmpN; i++) {
-            _outputs = _outputs.concat(_tx.extractOutputAtIndex(i));
-        }
+        bytes memory _proof = abi.encodePacked(_txid, _intermediateNodes, _merkleRoot);
+        // If the Merkle proof failed, bubble up error
+        return _proof.verifyHash256Merkle(_index);
     }
 
     /// @notice             Hashes transaction to get txid
-    /// @dev                This supports ONLY WITNESS INPUTS AND OUTPUTS
+    /// @dev                This supports legacy now
     /// @param _version     4-bytes version
-    /// @param _nInputs     1-byte num inputs
-    /// @param _inputs      Raw bytes inputs
-    /// @param _nOutputs    1-byte num outputs
-    /// @param _outputs     Raw bytes ouputs
+    /// @param _vin         Raw bytes length-prefixed input vector
+    /// @param _vout        Raw bytes length-prefixed output vector
+    /// @ param _locktime   4-byte tx locktime
     /// @return             32-byte transaction id, little endian
-    function transactionHash(
+    function calculateTxId(
         bytes memory _version,
-        bytes memory _nInputs,
-        bytes memory _inputs,
-        bytes memory _nOutputs,
-        bytes memory _outputs,
+        bytes memory _vin,
+        bytes memory _vout,
         bytes memory _locktime
-    ) public pure returns (bytes32) {
+    ) internal pure returns (bytes32) {
         // Get transaction hash dSha256(version + nIns + inputs + nOuts + outputs + locktime)
```

Comment:
Let's explain `dSha256` it a little more that it's double sha256 hash or `sha256((sha256(version + nIns + inputs + nOuts + outputs + locktime))`